packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -2968,10 +3006,41 @@ export function startHostTransition<F>(
     NoPendingHostTransition,
     // TODO: We can avoid this extra wrapper, somehow. Figure out layering
     // once more of this function is implemented.
-    () => callback(formData),
+    () => {
+      // Automatically reset the form when the action completes.
+      requestFormReset(formFiber);
+      return callback(formData);
+    },
   );
 }
 
+function requestFormReset(formFiber: Fiber) {
+  const transition = requestCurrentTransition();
+
+  if (__DEV__) {
+    if (transition === null) {
+      // An optimistic update occurred, but startTransition is not on the stack.
+      // The form reset will be scheduled at default (sync) priority, which
+      // is probably not what the user intended. Most likely because the
+      // requestFormReset call happened after an `await`.
+      // TODO: Theoretically, requestFormReset is still useful even for
+      // non-transition updates because it allows you to update defaultValue
+      // synchronously and then wait to reset until after the update commits.
+      // I've chosen to warn anyway because it's more likely the `await` mistake
+      // described above. But arguably we shouldn't.
+      console.error(
+        'requestFormReset was called outside a transition or action. To ' +
+          'fix, move to an action, or wrap with startTransition.',
+      );
+    }
+  }
+
+  const newResetState = {};
+  const resetStateHook: Hook = (formFiber.memoizedState.next: any);
+  const resetStateQueue = resetStateHook.queue;
+  dispatchSetState(formFiber, resetStateQueue, newResetState);
```

Comment:
Does this actually have to be on the Fiber?

Can't it just be scheduled on the Root with a list of forms to reset when the root commits?

Reply:
That way you don't need all the special commit phase stuff.

Reply:
Hmm I'll need to think more about that, the scenario I had in mind was if the form is inside a Suspense boundary, the root could commit before the input's defaultValue had been updated. Usually this shouldn't happen because during a Transition we won't hide already visible forms, but during a popstate transition we might.

The `<input form="myform>` case is interesting there, too.